File: 350/solution.go
package main

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"log"
	"strings"
)

const (
	layout        = "YDEyMzQ1Njc4OTAtPQpRV0VSVFlVSU9QW11cCkFTREZHSEpLTDsnCFpYQ1ZCTk0sLi8="
	shiftedLayout = "fiFAIyQlXiYqKClfKwpQe318Ckw6IgpNPD4/"
)

var globalLeftShiftMap map[byte]byte

func init() {
	globalLeftShiftMap = make(map[byte]byte)
	layouts := []string{layout, shiftedLayout}

	for _, layout := range layouts {
		decodedLayout, err := base64.StdEncoding.DecodeString(layout)
		if err != nil {
			log.Fatalf("decoding error: %s", err)
		}

		leftShiftMap := createLeftShiftMap(decodedLayout)
		globalLeftShiftMap = mergeMaps(globalLeftShiftMap, leftShiftMap)
	}
}

func main() {
	fmt.Println(translateRightShift(";p; epeor"))
	fmt.Println(translateRightShift("ejp s, o"))
	fmt.Println(translateRightShift("kpm nr;;opm od s hrmoid"))
}

func translateRightShift(input string) string {
	return translateString(globalLeftShiftMap, input)
}

func translateString(leftShiftMap map[byte]byte, input string) string {
	var result strings.Builder
	for _, c := range input {
		if value, ok := leftShiftMap[byte(c)]; ok {
			result.WriteByte(value)
		} else {
			result.WriteRune(c)
		}
	}
	return result.String()
}

// Creates an object that maps a key to the key that precedes it (key to its left).
func createLeftShiftMap(layout []byte) map[byte]byte {
	leftShiftMap := make(map[byte]byte)
	lowercased := bytes.ToLower(layout)
	layoutRows := bytes.Split(lowercased, []byte("\n"))
	for _, layoutRow := range layoutRows {
		leftShiftMap = mergeMaps(leftShiftMap, createLeftShiftMapForLayoutRow(layoutRow))
	}
	return leftShiftMap
}

// Creates an object that maps a key to the key that precedes it (key to its left).
func createLeftShiftMapForLayoutRow(layoutRow []byte) map[byte]byte {
	leftShiftMap := make(map[byte]byte)
	for i, c := range layoutRow {
		if i > 0 {
			leftShiftMap[c] = layoutRow[i-1]
		}
	}
	return leftShiftMap
}

// Merges b into a. Returns a.
func mergeMaps(a map[byte]byte, others ...map[byte]byte) map[byte]byte {
	for _, otherMap := range others {
		for k := range otherMap {
			a[k] = otherMap[k]
		}
	}
	return a
}

File: 350/solution.js
// const layout = `\`1234567890-=\nqwertyuiop[]\\\nasdfghjkl;'\nzxcvbnm,./`;
// const upperCaseLayout = `QWERTYUIOP\nASDFGHJKL\nZXCVBNM`;
// const shiftedLayout = `~!@#$%^&*()_+\np{}|\nl:"\nm<>?`;
const layout = "YDEyMzQ1Njc4OTAtPQpxd2VydHl1aW9wW11cCmFzZGZnaGprbDsnCnp4Y3Zibm0sLi8=";
const upperCaseLayout = "UVdFUlRZVUlPUApBU0RGR0hKS0wKWlhDVkJOTQ==";
const shiftedLayout = "fiFAIyQlXiYqKClfKwpwe318Cmw6IgptPD4/";
const leftShiftMaps = [layout, upperCaseLayout, shiftedLayout].map(atob).map(createLeftShiftMap);

main();

function main() {
  console.log(translateRightShift(";p; epeor"));
  console.log(translateRightShift("ejp s, o"));
  // Easter egg ^_^
  console.log(translateRightShift(`jyy[d"zzupiyi/nrzfWe5e0EhCvW`));
}

function translateRightShift(input) {
  return translateString(mergeObjects(...leftShiftMaps), input);
}

function translateString(leftShiftMap, input) {
  return Object.
    values(input).
    map(char => leftShiftMap[char] ?? char).
    join("");
}

/**
 * Creates an object that maps a key to the key that precedes it (key to its left).
 * @param {string} layout - a string of keys representing keyboard layout.
 * @returns {Record<string, string | undefined>}
 */
function createLeftShiftMap(layout) {
  return layout.
    split('\n').
    reduce((leftMap, layoutRow) => {
      return mergeObjects(leftMap, createLeftShiftMapForLayoutRow(layoutRow))
    }, {});
}

/**
 * Creates an object that maps a key to the key that precedes it (key to its left).
 * @param {string} layoutRow - a string of keys
 * @returns {Record<string, string | undefined>}
 */
function createLeftShiftMapForLayoutRow(layoutRow) {
  return Object.
    values(layoutRow).
    reduce((leftMap, char, i) => {
      leftMap[char] = i > 0 ? layoutRow[i - 1] : layoutRow[layoutRow.length - 1];
      return leftMap;
    }, {});
}

/**
 * Merges two objects in reverse order. Values in a replace values from b.
 * @param {object} a 
 * @param {object} b 
 * @returns {object}
 */
function mergeObjects(a, ...args) {
  if (args.length == 0) return a;
  return { ...args[args.length - 1], ...mergeObjects(a, ...args.slice(0, -1)) };
}